src/common/lib/api.ts

```diff
@@ -33,64 +29,76 @@ export interface AccountInfoRes {
   name: string;
 }
 
-interface StatusRes {
+export interface StatusRes {
   configured: boolean;
   unlocked: boolean;
   currentAccountId: string;
 }
 
-interface UnlockRes {
+export interface UnlockRes {
   currentAccountId: string;
 }
 
 interface BlocklistRes {
   blocked: boolean;
 }
 
-export const getAccountInfo = () => msg.request<AccountInfoRes>("accountInfo");
-
-/**
- * stale-while-revalidate get account info
- * @param id - account id
- * @param callback - will be called first with cached (stale) data first, then with fresh data.
- */
-export const swrGetAccountInfo = async (
-  id: string,
-  callback?: (account: AccountInfo) => void
-): Promise<AccountInfo> => {
-  const accountsCache = await getAccountsCache();
-
-  return new Promise((resolve, reject) => {
-    if (accountsCache[id]) {
-      if (callback) callback(accountsCache[id]);
-      resolve(accountsCache[id]);
-    }
-
-    // Update account info with most recent data, save to cache.
-    getAccountInfo()
-      .then((response) => {
-        const { alias } = response.info;
-        const { balance: resBalance, currency } = response.balance;
-        const name = response.name;
-        const balance =
-          typeof resBalance === "number" ? resBalance : parseInt(resBalance); // TODO: handle amounts
-        const account = {
-          id,
-          name,
-          alias,
-          balance,
-          currency: currency || "BTC", // set default currency for every account
-        };
-        storeAccounts({
-          ...accountsCache,
-          [id]: account,
-        });
-        if (callback) callback(account);
-        return resolve(account);
-      })
-      .catch(reject);
-  });
+export const getAccountInfo = (): Promise<AccountInfoRes> =>
+  msg.request<AccountInfoRes>("accountInfo");
+
+export const getAccountInfoKey = (
+  id: string | null | undefined
+): string | null => (id ? `accountInfo/${id}` : null);
+
+// useSWR hook is expected to be called in every render, hence we expect parameter id null to skip fetching
+export const useAccountInfoCached = async (
+  id: string | null
+): Promise<AccountInfo | null> => {
+  const { data, error } = await useSWR<AccountInfoRes>(
+    getAccountInfoKey(id),
+    getAccountInfo
+  );
+
+  if (!data) return null;
+  // @Todo: how to handle this error in the calling function ?
+  if (error) throw error;
+
+  return buildAccountInfo(id, data);
 };
+
+// revalidation (mark the data as expired and trigger a refetch) for the resource
+export const mutateAccountInfo = async (
+  id: string | null
+): Promise<AccountInfo | null> => {
+  const data = await mutate<AccountInfoRes>(
+    getAccountInfoKey(id),
+    getAccountInfo
+  );
+  if (!data) return null;
+
+  return buildAccountInfo(id, data);
+};
+
+const buildAccountInfo = (
+  id: string | null,
```

Comment:
the id should exist, I don't think it should be nullable here. I think we should never use buildAccountInfo if the id or data isn't available.

Reply:
This is actually due to a Typescript type problem: due to the id coming from the calling function "useAccountInfoCached" that calls the "useSWR" hook. By design the hooks need to be called every single render turn and the mechanics to skip it is to pass a key of null to useSWR. That's why the parent function id parameter can be null and the type here therefore as well. Do you have an idea how to solve it? If I try non-null assertion "id!" with the id argument, I get the lint error: "Forbidden non-null assertion  @typescript-eslint/no-non-null-assertion"

Reply:
Actually I'm using "// eslint-disable-next-line @typescript-eslint/no-non-null-assertion" in this function. I can move it up to the calling function. But the issue of an id parameter of null coming in will stay there, because of what I explained above. Unless you have a better idea.